cmudb/extensions/tscout/main.c

```diff
@@ -0,0 +1,224 @@
+                                   const char *queryString, ParamListInfo params, QueryEnvironment *queryEnv);
+static void WalkPlan(Plan *plan, ExplainState *es);
+static void explainXs(Plan *node, ExplainState *es);
+static size_t get_field_size(c_type type);
+
+static ExplainOneQuery_hook_type chain_ExplainOneQuery = NULL;
+
+void _PG_init(void) {
+  elog(LOG, "Initializing extension");
+
+  // Chain the Extension Explain Query Wrapper to the head of the chain.
+  chain_ExplainOneQuery = ExplainOneQuery_hook;
+  ExplainOneQuery_hook = ExplainOneQueryWrapper;
+
+  // Init logic (like setting flags) go here
+}
+
+void _PG_fini(void) {
+  // Clean up logic goes here
+  ExplainOneQuery_hook = chain_ExplainOneQuery;
+  elog(DEBUG1, "Finishing extension");
+}
+
+static void ExplainOneQueryWrapper(Query *query, int cursorOptions, IntoClause *into, ExplainState *es,
+                                   const char *queryString, ParamListInfo params, QueryEnvironment *queryEnv) {
+  PlannedStmt *plan;
+  QueryDesc *queryDesc;
+  instr_time plan_start, plan_duration;
+  int eflags = 0;
+
+  if (chain_ExplainOneQuery) {
+    chain_ExplainOneQuery(query, cursorOptions, into, es, queryString, params, queryEnv);
+  }
+
+  // Postgres does not expose an interface to call into the standard ExplainOneQuery.
+  // Hence, we duplicate the operations performed by the standard ExplainOneQuery ie
+  // calling into the standard planner.
+  // A non-standard planner can be hooked in, in the the future.
+  INSTR_TIME_SET_CURRENT(plan_start);
+  plan = (planner_hook ? planner_hook(query, queryString, cursorOptions, params)
+                       : standard_planner(query, queryString, cursorOptions, params));
+
+  INSTR_TIME_SET_CURRENT(plan_duration);
+  INSTR_TIME_SUBTRACT(plan_duration, plan_start);
+
+  TS_MARKER("TS_MARKER <NAME>");
+
+  if (es->format == EXPLAIN_FORMAT_TSCOUT) {
+    queryDesc =
+        CreateQueryDesc(plan, queryString, InvalidSnapshot, InvalidSnapshot, None_Receiver, params, queryEnv, 0);
+
+    if (es->analyze)
+      eflags = 0;
+    else
+      eflags = EXEC_FLAG_EXPLAIN_ONLY;
+    if (into) eflags |= GetIntoRelEFlags(into);
+
+    // Run the executor
+    ExecutorStart(queryDesc, eflags);
+    // This calls into initPlan() which populates the plan tree.
+    // TODO: Create a hook to executor start.
+
+    // Finally, walk through the plan, dumping its out in a separate top-level group.
+    ExplainOpenGroup("TscoutProps", NULL, true, es);
+    WalkPlan(queryDesc->planstate->plan, es);
+    ExplainCloseGroup("TscoutProps", NULL, true, es);
+
+    // Free the created query description resources.
+    ExecutorEnd(queryDesc);
+    FreeQueryDesc(queryDesc);
+  }
+
+  // Finally, after performing the extension specific operations, explain the planner.
+  ExplainOnePlan(plan, into, es, queryString, params, queryEnv, &plan_duration, NULL);
+}
+
+/**
+ * @brief Fetches the size of the field
```

Comment:
Style nit: please describe a function from the perspective of the caller. Suppose your caller has:

```
// comment
get_field_size(T_BOOL);
```

Try replacing "comment" with "Fetch the size of the field" vs "Fetches the size of the field".

```suggestion
 * @brief Fetch the size of the field.
```